crates/lsp/src/logging.rs

```diff
@@ -0,0 +1,324 @@
+// --- source
+// authors = ["Charlie Marsh"]
+// license = "MIT"
+// origin = "https://github.com/astral-sh/ruff/blob/03fb2e5ac1481e498f84474800b42a966e9843e1/crates/ruff_server/src/trace.rs"
+// ---
+
+//! The logging system for `air lsp`.
+//!
+//! ## Air crate logs
+//!
+//! For air crates, a single log level is supplied as one of: error, warn, info, debug,
+//! or trace, which is applied to all air crates that log.
+//!
+//! Resolution strategy:
+//!
+//! - The environment variable `AIR_LOG_LEVEL` is consulted.
+//!
+//! - The LSP `InitializeParams.initializationOptions.logLevel` option is consulted. This
+//!   can be set in VS Code / Positron using `air.logLevel`, or in Zed by supplying
+//!   `initialization_options`.
+//!
+//! - If neither are supplied, we fallback to `"info"`.
+//!
+//! ## Dependency crate logs
+//!
+//! For dependency crates, either a single log level can be supplied, or comma separated
+//! `target=level` pairs can be supplied, like `tower_lsp=debug,tokio=info`.
+//!
+//! Resolution strategy:
+//!
+//! - The environment variable `AIR_DEPENDENCY_LOG_LEVELS` is consulted.
+//!
+//! - The LSP `InitializeParams.initializationOptions.dependencyLogLevels` option is
+//!   consulted. This can be set in VS Code / Positron using `air.dependencyLogLevel`, or
+//!   in Zed by supplying `initialization_options`.
+//!
+//! - If neither are supplied, we fallback to no logging for dependency crates.
+//!
+//! ## IDE support
+//!
+//! For VS Code and Zed, which are known to support `window/logMessage` well, logging will
+//! emit a `window/logMessage` message. Otherwise, logging will write to `stderr`,
+//! which should appear in the logs for most LSP clients.
+use core::str;
+use serde::Deserialize;
+use std::fmt::Display;
+use std::io::{Error as IoError, ErrorKind, Write};
+use std::str::FromStr;
+use tokio::sync::mpsc::unbounded_channel;
+use tower_lsp::lsp_types::ClientInfo;
+use tower_lsp::lsp_types::MessageType;
+use tower_lsp::Client;
+use tracing_subscriber::filter;
+use tracing_subscriber::fmt::time::LocalTime;
+use tracing_subscriber::fmt::TestWriter;
+use tracing_subscriber::{
+    fmt::{writer::BoxMakeWriter, MakeWriter},
+    layer::SubscriberExt,
+    Layer,
+};
+
+// TODO:
+// - Add `air.logLevel` and `air.dependencyLogLevels` as VS Code extension options that set
```

Comment:
I'm also considering going the rust-analyzer route of having lots of smaller namespaces for options, like in this case:

```
air.log.level
air.log.dependencyLevels
```

which would come in as a `LogSettings` struct containing `level` and `dependencyLevels` fields.

That seems like it would be useful for grouping related options?